### PR TITLE
Use QtWebEngine instead of deprecated QtWebView

### PR DESCRIPTION
**Pull request by katafrakt**
As mentioned in #54, QtWebKit was deprecated and is now removed from Qt releases. Although there is apparently some effort being made to reinstate it, right now it results in complicated build process to get Manuscript working (see #35). And since it uses QtWebKit only to render HTML and PDF previews, it is fairly easy to rewrite it using QtWebKit's successor: QtWebEngine.

This PR makes it.

However, note that I only tested it with HTML preview now (it worked). I need to fix my LaTeX installation to test it with PDFs. Also, any clues on testing would be most welcome.

With QtWebEngine instead of QtWebKit, it's possible to build Manuscript against standard binary distribution of Qt from Homebrew and I suppose it's also easier on some Linux distributions. I don't have a chance right now to check it on Windows.

### CODE REVIEW

**Comment by katafrakt** [comment]
One bug found. When I tried to get a preview of sample project on Linux, first I get:

```
! Package inputenc Error: Unicode char ✡ (U+2721)
(inputenc)                not set up for use with LaTeX.

See the inputenc package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.248 among the people.'✡

Try running pandoc with --latex-engine=xelatex.
pandoc: Error producing PDF
```

Why? It does not seem Qt-related?

But then it results in:
```
Traceback (most recent call last):
  File "/home/katafrakt/dev/github/manuskript/bin/../manuskript/ui/exporters/exporter.py", line 90, in preview
    F.preview(self.settingsWidget, self.previewWidget)
  File "/home/katafrakt/dev/github/manuskript/bin/../manuskript/exporter/pandoc/PDF.py", line 52, in preview
    previewWidget.loadPDF(filename)
  File "/home/katafrakt/dev/github/manuskript/bin/../manuskript/ui/views/PDFViewer.py", line 20, in loadPDF
    self.settings.clearMemoryCaches()
AttributeError: 'QWebEngineSettings' object has no attribute 'clearMemoryCaches'
Fatal Python error: Aborted
```

And yes, `QWebEngineSettings` does not have `clearMemoryCaches` attribute. Investigation needed if this can be skipped or how to replace it.

**Comment by gedakc** [comment]
Following are the results of testing this PR on the following platforms:

a) Windows XP with Python 3.4.4 and PyQt v5.5.1
b) Kubuntu 16.04 with Python 3.5.2 and PyQt v5.5.1+dfsg-3ubuntu4

Unfortunately manuskript failed to run on both of the above platforms.  ;-(

The error message:
```
gedakc@kaby:~/workspace/manuskript$ manuskript
Traceback (most recent call last):
  File "/home/gedakc/workspace/manuskript/bin/manuskript", line 11, in <module>
    from manuskript import main
  File "/home/gedakc/workspace/manuskript/bin/../manuskript/main.py", line 7, in <module>
    import PyQt5.QtWebEngineWidgets # must be imported before QApplication
ImportError: No module named 'PyQt5.QtWebEngineWidgets'
```

So far I have been unable to find QtWebEngine packages for the above versions.

I also discovered the following message on [Deploying Qt WebEngine Applications](https://doc.qt.io/qt-5/qtwebengine-deploying.html).

> On Windows, QtWebEngine only supports Windows Vista or newer as target platform. Due to use of newer API in Chromium, **Windows XP is not supported**. WinRT is not supported, either.

Based on these findings I think if we are to implement QtWebEngine then we will need a solution that can also continue to use QtWebView.

Perhaps there is a way with Python to conditionally use one or the other?

**Comment by olivierkes** [comment]
The [unicode bug](https://github.com/olivierkes/manuskript/issues/117) is fixed.

I could not install QtWebEngine, so I wasn't able to test your PR. But using your code, I tried a solution so that manuskript can use either QtWebKit, QtWebEngine, or none if none is available. 

Would you mind testing it, both of you?
Its on the [`feature/QtWebEngine`](https://github.com/olivierkes/manuskript/tree/feature/QtWebEngine) branch.

Thanks.

**Comment by olivierkes** [comment]
Hop, I realised I could commit on your branch in your repo. Great !

I commented out the import before the qApplication because I could find no references to that. Where do you see this as necessary? That seems strange to me.

**Comment by gedakc** [comment]
# TLDR - To Long Didn't Read;

The _feature/QtWebEngine_ branch works when the _QtWebKit_ is available.
No testing was done with _QtWebEngine_.

Hopefully @katafrakt can provide results from testing with **QtWebEngine**.


# PREPARATION

On *buntu ensure:

- Packages pandoc and texlive-xetex (provides `xelatex`) are installed:  
  `sudo apt-get install pandoc texlive-xetex`

- Package python3-markdown is installed:
  `sudo apt-get install python3-markdown`

- Manuskript exporter for Pandoc set _LaTeX engine used to produce the
  PDF._ to **xelatex**.
  See issue #117.


On Windows XP ensure:

- Package for python3 markdown is installed:
  ```
  C:
  cd \Python34\Scripts
  pip install markdown
  ```

- Windows XP has neither Pandoc nor Latex installed.


# TESTING

## Environments

```
    OS        | Python | PyQt                | QtWebKit | QtWebEngine
--------------+------------------------------+----------+------------
Kubuntu 16.04 | 3.5.2  | 5.5.1+dfsg-3ubuntu4 | Included | Not incl.
Ubuntu  17.04 | 3.5.3  | 5.7+dfsg-5          | Included | Not incl.
Windows XP    | 3.4.4  | 5.5.1               | Included | Not incl.
```


## Steps

The steps I used for testing the feature/QtWebEngine branch are as follows:

1.  Open Manuskript with a project that contains some markdown content.
2.  Select **File -> Compile**.
3.  Set _Export to_ **HTML**.
4.  Click **Preview** button in _Export_ window.
5.  TEST A:  Review contents of _HTML Output_ tab.
6.  Set _Export to_ **Pandoc - HTML**.
7.  Click **Preview** button in _Export_ window.
8.  TEST B:  Review contents of _HTML Output_ tab.
9.  Set _Export to_ **Pandoc - PDF**.
10. Click **Preview** button in _Export_ window.
11. TEST C:  Review contents of preview pane.
12. Click **Export** button.
13. Follow prompts to export to a PDF file (e.g., file.pdf).
14. TEST D: Review PDF file file.pdf contents in a PDF viewer.
15. Close _Export_ window.


## RESULTS

```
              |    P R E V I E W S                  | F I L E
              | TEST A   | TEST B      | TEST C     | TEST D
    OS        | HTML     | Pandoc HTML | Pandoc PDF | PDF File
--------------+----------+--------------+-----------+-----------
Kubuntu 16.04 | Works    | Works       | Empty [1]  | Works
Ubuntu  17.04 | Works    | Works       | Empty [1]  | Works
Windows XP    | Works    | Not Avail.  | Not avail. | Not avail.
```

[1] Blank / empty preview screen shown.  No story contents visible.


NOTE: None of the above platforms had QtWebEngine installed.


# SUMMARY

Based on my testing, the _feature/QtWebEngine_ branch works when the
_QtWebKit_ is available.  No testing was done with _QtWebEngine_.

Hopefully @katafrakt can provide results from testing with **QtWebEngine**.

**Comment by olivierkes** [comment]
Great report! From what I read, it does not exactly work when webkit is installed, because preview doesn't work, though it should. Is that right?

**Comment by gedakc** [comment]
@olivierkes that is correct.  The preview pane is blank (in fact it shows a black screen) when I select **Pandoc - PDF**.

Regarding _QtWebEngine_, I will see if I can find a GNU/Linux distributions that has packages for _QtWebEngine_ so I can test this code path as well.

**Comment by gedakc** [comment]
It turns out that Ubuntu 17.04 has both _python-pyqt5.qtwebkit_ and _python-pyqt5.qtwebengine_ packages.  Hence to test I removed qtwebkit and installed qtwebengine.

See the updated testing results that are posted earlier in this thread.

Hopefully @katafrakt can provide results from testing with QtWebEngine on Mac OS X.

**Comment by katafrakt** [comment]
I'll test MacOS + arch (it worked for me with QtWebEngine before) later today or tomorrow and will get back to you.

**Comment by gedakc** [comment]
When running the _features/QtWebEngine_ branch with TEST B (preview of Pandoc HTML) the following messages appear in the terminal window:

```
Running manuskript version 0.4.0.
Note: No translator found or loaded for locale en_CA.
Qt WebEngine seems to be initialized from a plugin. Please set Qt::AA_ShareOpenGLContexts using QCoreApplication::setAttribute before constructing QGuiApplication.
Debug: Web rendering engine used: QTextEdit
Loading: /home/user/manuskript-test-one.msk
Detected file format version: 1. Zip: False.
Project /home/user/manuskript-test-one.msk loaded.
Traceback (most recent call last):
  File "/home/user/manuskript/bin/../manuskript/ui/exporters/exporter.py", line 90, in preview
    F.preview(self.settingsWidget, self.previewWidget)
  File "/home/user/manuskript/bin/../manuskript/exporter/pandoc/HTML.py", line 43, in preview
    previewWidget.widget(2).setHtml(html, QUrl.fromLocalFile(path))
TypeError: setHtml(self, str): too many arguments
```

'Looks like we might need a different HTML call for QtWebEngine....

**Comment by katafrakt** [comment]
As for this question:

> I commented out the import before the qApplication because I could find no references to that. Where do you see this as necessary? That seems strange to me.

Without that I'm getting this error:

```
    from PyQt5.QtWebEngineWidgets import QWebEngineView
ImportError: QtWebEngineWidgets must be imported before a QCoreApplication instance is created
```

Here's possible explanation: https://riverbankcomputing.com/pipermail/pyqt/2016-August/037928.html

**Comment by katafrakt** [comment]
@gedakc Regarding your last comment - the listing you pasted says `Debug: Web rendering engine used: QTextEdit` so it seems QtWebEngine has not been detected or loaded.

**Comment by katafrakt** [comment]
Here are results of my tests:

| OS | Engine | Test A<br/>HTML | Test B<br/>Pandoc HTML | Test C<br/>Pandoc PDF preview | Test D <br />Pandoc PDF export |
|-|-|-|-|-|-|
| MacOS | QtWebEngine | works | works | blank [1] | works |
| Arch Linux | QtWebEngine | works | works | blank [1] | works |
| Arch Linux | QtWebKit | works | works | blank [2] | works |

### Notes

[1] Empty container for PDF file is visible, message in console says:
```
js: Uncaught TypeError: Cannot read property 'PDFJS' of undefined
```

[2] Same as above, but without any message

empty_pdf

In both cases, I think I'm missing something obvious to get it to work.

#### Other notes

I tested both `feature/QtWebEngine` and `qtwebengine` (for this PR) branches, but the results were exactly the same, so I skipped the difference.

I have to reenable early import of `PyQt5.QtWebEngine` to avoid errors mentioned in my [previous comment](https://github.com/olivierkes/manuskript/pull/116#issuecomment-309211570).

To test on Linux, I had to alter `webView.py` file, so I can tell Manuskript which engine I want to use explicitly. File after changes is:

```python
#!/usr/bin/env python
# --!-- coding: utf8 --!--
import PyQt5
import os

features = {'qtwebkit': False, 'qtwebengine': False}

if 'QT_WEB' in os.environ:
    features[os.environ['QT_WEB']] = True
else:
    try:
        import PyQt5.QtWebKitWidgets
        features['qtwebkit'] = True
    except:
        features['qtwebkit'] = False

    try:
        import PyQt5.QtWebEngineWidgets
        features['qtwebengine'] = True
    except:
        features['qtwebengine'] = False

if features['qtwebkit']:
    from PyQt5.QtWebKitWidgets import QWebView
    print("Debug: Web rendering engine used: QWebView")
    webEngine = "QtWebKit"
    webView = QWebView
elif features['qtwebengine']:
    from PyQt5.QtWebEngineWidgets import QWebEngineView
    print("Debug: Web rendering engine used: QWebEngineView")
    webEngine = "QtWebEngine"
    webView = QWebEngineView
else:
    from PyQt5.QtWidgets import QTextEdit
    print("Debug: Web rendering engine used: QTextEdit")
    webEngine = "QTextEdit"
    webView = QTextEdit
```

Hope this helps.

**Comment by gedakc** [comment]
Thanks @katafrakt for testing and for the tip that QtWebEngineWidgets must be imported before a QCoreApplication instance is created.

I was successfully able to view a TEST B **Pandoc - HTML** preview on Ubuntu 17.04 after adding the following line:

```
$ git diff
diff --git a/manuskript/main.py b/manuskript/main.py
index df32153..7da361b 100644
--- a/manuskript/main.py
+++ b/manuskript/main.py
@@ -4,6 +4,7 @@ import faulthandler
 import os
 import sys
 
+import PyQt5.QtWebEngineWidgets # must be imported before QApplication
 from PyQt5.QtCore import QLocale, QTranslator, QSettings
 from PyQt5.QtGui import QIcon
 from PyQt5.QtWidgets import QApplication, qApp
```

Note that we'll need to do a conditional import in the code so that both QtWebKit and QtWebEngine will work.

**Comment by katafrakt** [comment]
> Note that we'll need to do a conditional import in the code so that both QtWebKit and QtWebEngine will work.

Sure thing. I'd love to see it as part of some global `Config` object or something like that, not just global variables lying around 😉

**Comment by gedakc** [comment]
> > Note that we'll need to do a conditional import in the code so that both QtWebKit and QtWebEngine will work.

> Sure thing. I'd love to see it as part of some global Config object or something like that, not just global variables lying around :wink:

My preference is to follow the KISS philosophy, and simply try to load one first, and if that fails then try to load the other.  That way we don't burden our users with trying to figure out why they should choose one over the other.  In fact they might not even have a choice if QtWebKit is left out of new PyQt packages.

Is there an advantage of using one versus the other? 

If so, then we could try to load the advantageous one first, whichever that might be:  QtWebKit or QtWebEngine.

**Comment by gedakc** [comment]
I've created a new [feature/QtWebEngine-v2](https://github.com/gedakc/manuskript/tree/feature/QtWebEngine-v2) branch that conditionally loads QtWebKitWidgets and QtWebEngineWidgets in the same order as used in manuskript/ui/views/webView.py.

Using this branch I have the same testing results as @katafrakt noted 4 posts earlier.

On Ubuntu 17.04 with QtWebEngine when trying TEST C **Pandoc -PDF** preview, I also see a blank pane in manuskript, and see the following message in my terminal window:

`js: Uncaught TypeError: Cannot read property 'PDFJS' of undefined`

It seems we are missing something needed by QtWebEngine for a PDF preview.

It is possible that the same "_missing something_" might make PDF preview work for QtWebKit as well.

**Comment by gedakc** [comment]
> `js: Uncaught TypeError: Cannot read property 'PDFJS' of undefined`

I came across the following mention of this issue:  [pdfjs-dist 1.3.94 throwing error](https://github.com/mozilla/pdf.js/issues/6760)

It appears that the work-around implemented was to roll back to an earlier version.

Because manuskript has `libs/pdf.js/...` included in the code, perhaps we could also try rolling back to an earlier version?

**Comment by olivierkes** [comment]
> I tested both feature/QtWebEngine and qtwebengine (for this PR) branches, but the results were exactly the same, so I skipped the difference.

PDF preview works for me in the current `develop` branch, as well as in `feature/QtWebEngine`, but does not work in the `qtwebengine` in @katafrakt's clone. Haven't been able to pinpoint why.

Can you please confirm that PDF preview doesn't work with `develop` or `feature/QtWebEngine`?

> Because manuskript has libs/pdf.js/... included in the code, perhaps we could also try rolling back to an earlier version?

I cannot reproduce `js: Uncaught TypeError: Cannot read property 'PDFJS' of undefined`, and the current version works for me sometimes. But please do try and tell us the results.

>> Note that we'll need to do a conditional import in the code so that both QtWebKit and QtWebEngine will work.

> Sure thing. I'd love to see it as part of some global Config object or something like that, not just global variables lying around :wink:

Not sure what you mean by that, sorry ;) For now, I updated `feature/QtWebEngine` for the conditional import. I simply `import manuskript.ui.views.webView` before creating the qApp, so it should import `QtWebEngineWidgets` if available.

I cannot find an easy way to install qtwebengine on Linux Mint 18.1, so I'm working blind here.

If that works, then we can can maybe merge that, and open a bug report for the PDF preview issue, which seems unconnected.

**Comment by gedakc** [comment]
Following are my recent test results using @olivierkes _develop_ and _feature/QtWebEngine_ branches.

| Branch | Engine | Test A<br/>HTML | Test B<br/>Pandoc HTML | Test C<br/>Pandoc PDF preview | Test D <br />Pandoc PDF export |
|-|-|-|-|-|-|
| develop | QtWebKit | works | works | blank [1] | works |
| develop | QtWebEngine | fails [2] | fails | fails | fails |
| feature/QtWebEngine | QtWebKit | works | works | blank [1] | works |
| feature/QtWebEngine | QtWebEngine | works | works | blank [3] | works |

[1] Preview pane contains blank / empty screen. No console error message.

[2] Manuskript fails to run because no QtWebKit installed.  Only QtWebEngine installed.

    <snip>
    ImportError: No module named 'PyQt5.QtWebKitWidgets'
    <snip>

[3] Preview pane contains blank / empty screen. Console has error message:

    js: Uncaught TypeError: Cannot read property 'PDFJS' of undefined

@olivierkes, for testing you might consider setting up a Virtual Machine with a newer OS that has the QtWebEngine package.  My guess is that you'll need to have at least 4 GB RAM and 20+ GB free space in a file system to run a VM.  On Ubuntu/Mint, you can install [VirtualBox](https://www.virtualbox.org/) with:

    sudo apt install virtualbox

Then you could install an OS familar to you with steps similar to the following:
[How to install Ubuntu on VirtualBox?](https://askubuntu.com/questions/142549/how-to-install-ubuntu-on-virtualbox)

Next I plan to investigate using an older version of _pdfjs_.

**Comment by gedakc** [comment]
I believe I have discovered the reason why Test C **Pandoc PDF** preview does not work for @katafrakt  or me.  It is because the `manuskript/libs/pdf.js/build` directory is missing.

The .gitignore file contains an entry for `build` and I suspect this is the reason why the   `manuskript/libs/pdf.js/build` directory was not included when @olivierkes  added the pdf.js library components.

To fix this issue I think if @olivierkes  adds this missing `build` directory to the _feature/QtWebEngine_ branch, then all of our tests should work on all of our platforms.

To commit the `build` directory I had to use the "-f" git commit parameter.

    git commit -f manuskript/libs/pdf.js/build/*


**BACKGROUND RESEARCH**

I tested several versions of [pdf.js](https://github.com/mozilla/pdf.js) including the latest stable release [pdfjs-1.7.225](https://github.com/mozilla/pdf.js/releases/tag/v1.7.225).

All of these versions worked for all features including Test C **Pandoc PDF** preview.

When comparing these versions to @olivierkes code, I noticed that the _feature/QtWebEngine_ branch was missing the `manuskript/libs/pdf.js/build` directory which should contain `pdf.js` and `pdf.worker.js` files.

**Comment by olivierkes** [comment]
> To fix this issue I think if @olivierkes adds this missing `build` directory to the *feature/QtWebEngine* branch, then all of our tests should work on all of our platforms.

Of course !! Thanks for finding that. See the updated branch now, with missing files, including license. Everything should work now I guess.

**Comment by gedakc** [comment]
Thank you @olivierkes for committing the missing files/directories.

Following are my latest test results using the _feature/QtWebEngine_ branch.

| OS | Engine | Test A<br/>HTML | Test B<br/>Pandoc HTML | Test C<br/>Pandoc PDF preview | Test D <br />Pandoc PDF export |
|-|-|-|-|-|-|
| Kubuntu 16.04 | QtWebKit | works | works | works | works |
| Ubuntu 17.04 | QtWebEngine | works | works | works [1] | works |

[1] PDF Preview works.  However many messages displayed on console.

Example of 198 lines of console messages:

```
js: #thumbs is undefined.
js: #thumbs is undefined.
js: #thumbs_label is undefined.
js: #thumbs_label is undefined.
js: #outline is undefined.
js: #outline is undefined.
js: #outline_label is undefined.
js: #outline_label is undefined.
js: #attachments is undefined.
js: #attachments is undefined.
<snip>
js: #document_properties_page_count is undefined.
js: #document_properties_page_count is undefined.
js: #document_properties_close is undefined.
js: #document_properties_close is undefined.
js: The provided value 'moz-chunked-arraybuffer' is not a valid enum value of type XMLHttpRequestResponseType.
js: #page_of is undefined.
js: #thumb_page_title is undefined.
js: #thumb_page_title is undefined.
```

I can provide a complete list of the console messages if needed.

**Comment by olivierkes** [comment]
No thanks, I guess it's unnecessary. There might be a way to turn those
warnings off. I'll check that later on.

**Comment by gedakc** [comment]
@katafrakt, does the updated _features/QtWebEngine_ branch work for you on MacOS and/or Arch Linux?

**Comment by katafrakt** [comment]
@olivierkes, @gedakc Hello again! First of all I'm terribly sorry for having abandoned this topic for such a long time. This was cause by my laptop breakdown at first and then I simply forgot about this issue.

Anyway, I retested `feature/QtWebEngine` branch on fresh Arch Linux installation and everything from our little test plan is working without any issues in the console etc. My branch also works, with addition of importing `QWebEngineView` before creating `QCoreApplication`.

I also tested on Mac, but only `feature/QtWebEngine` with `QtWebEngine`. Everything works there too.

I would like to suggest to proceed: choose which branch to clean up and merge it, maybe release. Then perhaps we can start working on packaging Manuskript for Mac? I think I like code in `feature/QtWebEngine` better, so why not merge it and close this PR? What do you think?

**Comment by gedakc** [comment]
Thank you @katafrakt for testing and posting your results.

I am in agreement to proceed with the code in `feature/QtWebEngine` and @olivierkes can clean it up if needed.  Then this PR can be close

I am favour including other open PRs as well and then creating new release.  With @katafrakt offering to build the MacOS image, then I offer to build the Windows and Linux PyInstaller images.

**Comment by katafrakt** [comment]
Just to be clear: I have no experience in building MacOS releases. I only briefly researched the topic and it seem not impossible. I volunteer to look more into it, however assumption that I will succeed may be a bit far-fetched :wink:

**Comment by gedakc** [comment]
@katafrakt, we have some instructions on building packages in the [Manuskript Wiki](https://github.com/olivierkes/manuskript/wiki) that may help.

**Comment by katafrakt** [comment]
I tried that yesterday and today, but with no luck. It seems it needs some tweaking. I'll let you know if I find a solution.

**Comment by gedakc** [comment]
I've had good results using PyInstaller on both Linux and Windows.  PyInstaller is listed as being supported on Mac OS X as well.

**Comment by olivierkes** [comment]
I merged the `feature/QtWebEngine` branch from into `develop`.

Is it worth mergin @gedakc's `feature/QtWebEngine-v3` branch? What are the differences between the two pdfjs versions?

**Comment by gedakc** [comment]
@olivierkes, please do not merge my _feature/QtWebEngine-v3_ branch.

It has been long enough since I created the branch that I no longer remember the differences.  I will remove my _feature/QtWebEngine-v2_ and _feature/QtWebEngine-v3_ branches.

**Comment by gedakc** [comment]
@katafrakt does the _features/QtWebEngine_ code that is committed to the _develop_ branch work for you?

If so then we should probably close this PR.

**Comment by katafrakt** [comment]
@gedakc I have just done a re-check. There are some problems with PDF preview because I apparently broke TeX installation on my computer, but generally things seem to work. I only needed to set pandoc PDF engine to xelatex manually.

I'm closing the PR.

**Comment by highwindmx** [comment]
> 
> 
> I merged the `feature/QtWebEngine` branch from into `develop`.
> 
> Is it worth mergin @gedakc's `feature/QtWebEngine-v3` branch? What are the differences between the two pdfjs versions?



> turn those warnings off.

may I ask how to suppress these warnings? :)
